aiomysql/connection.py

```
@@ -255,35 +258,38 @@ def closed(self):
         """The readonly property that returns ``True`` if connections is
         closed.
         """
-        return self._writer is None
+        return self._protocol is None
 
     @property
     def encoding(self):
         """Encoding employed for this connection."""
-        return self._encoding
+        return self._protocol.encoding
 
     @property
     def charset(self):
         """Returns the character set for current connection."""
-        return self._charset
+        return self._protocol._charset
 
     def close(self):
         """Close socket connection"""
-        if self._writer:
-            self._writer.transport.close()
-        self._writer = None
-        self._reader = None
+        if self._protocol is not None and not self._protocol.closed and \
+                self._protocol.transport is not None:
+            try:
+                self._protocol.transport.close()
+            except Exception:
```

do not think suppressing exception is good idea